### PR TITLE
Wip/mehrad/updates

### PR DESCRIPTION
**Pull request by Mehrad0711**
"Fixing requote.js errors"
Fixing a couple of bugs when requoting consecutive tokens or parameters with lexical overlap.
It will now consume almost anything as long as the sentence and the program have matching parameters!

### CODE REVIEW

**Comment by gcampax** [comment]
What's the bug in requote.js specifically? Do you have a test case that fails with master? I've run requote.js extensively, and also fixed it a bit recently (hence the merge conflict). I did not need the complete rewrite you did.

**Comment by Mehrad0711** [comment]
Belive me you do! There are a lot of cases when the script fails. Just run it once on the restaurants trains set: s3://almond-research/mehrad/dataset/multilanguage/restaurants/final1-original-en/train.tsv
Example: 
```RPS5000099412-1	what are restaurant with their summary containing " and may not " but not " and " and " italian " in the served cuisine ?	now => ( @org.schema.Restaurant.Restaurant ) filter not param:description:String =~ " and " and param:description:String =~ " and may not " and param:servesCuisine:String =~ " italian " => notify```

File: tool/requote.js [hunk 1]
```diff
@@ -17,22 +17,35 @@ const { DatasetParser, DatasetStringifier } = require('../lib/dataset-parsers');
...
         let found = true;
+
         for (let j = 0; j < substring.length; j++) {
             if (sequence[i+j] !== substring[j]) {
                 found = false;
                 break;
             }
         }
-        if (found)
+        if (found && check_range(i, spansBySentencePos))
```
Comment (gcampax) [inline tool/requote.js]: This (non-overlapping spans) and the sorting by span length are the only functional changes, correct? The rest is just code cleanup?

Comment (Mehrad0711) [reply tool/requote.js]: yeah. I found the bug. It was not correctly reverting the position orders. it should be fixed now.

**Comment by Mehrad0711** [comment]
But yeah we need more comprehensive unit tests. There are a bunch of corner cases missing.

File: test/index.js [hunk 2]
```diff
@@ -16,17 +16,22 @@ process.env.TEST_MODE = '1';
...
-seq([
-    ('./test_stream_utils'),
-    ('./test_requoting'),
-    ('./test_sentence_generator'),
+
+// test lib scripts
+do_test([
+    // ('./test_stream_utils'),
+    // ('./test_requoting'),
+    // ('./test_sentence_generator'),
```
Comment (gcampax) [inline test/index.js]: Why the changes here?

File: test/everything.sh [hunk 3]
```diff
@@ -5,6 +5,8 @@ set -x
 
 srcdir=`dirname $0`/..
 
+$srcdir/test/test_tools.sh
```
Comment (gcampax) [inline test/everything.sh]: You don't need another shell file. Either put it as unit test (in index.js) or as functional test in cmdline.sh

File: test/test_tools.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+on_error() {
+    rm -fr ./test_results_temp/
+}
+trap on_error ERR INT TERM
+
+mkdir -p ./test_results_temp/
+
+# test requote.js
+$(which node) $srcdir/tool/genie.js requote $srcdir/test/data/samples-en-hard.tsv --output ./test_results_temp/samples-en-hard-requoted.tsv --mode replace
+cmp --silent ./test_results_temp/samples-en-hard-requoted.tsv $srcdir/test/data/samples-en-hard-requoted.tsv && echo '*** SUCCESS: Test passed! ***' || { echo '*** ERROR: Test has Failed... ***'; exit 1 ;}
```
Comment (gcampax) [inline test/test_tools.sh]: Please be consistent with other tests: use `diff`, be silent on success, and show the diff on error.

File: test/test_tools.sh [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+set -x
+
+srcdir=`dirname $0`/..
+
+on_error() {
+    rm -fr ./test_results_temp/
+}
+trap on_error ERR INT TERM
+
+mkdir -p ./test_results_temp/
```
Comment (gcampax) [inline test/test_tools.sh]: Tests should execute in a temporary directory, otherwise they'll clobber each other on multiple executions.

File: tool/requote.js [hunk 6]
```diff
@@ -273,16 +271,16 @@ function requoteSentence(id, sentence, program, mode) {
     sentence = sentence.split(' ');
     program = program.split(' ');
 
-    let [spansBySentencePos, spansByProgramPos_sorted, sortIndices] = findSpanPositions(id, sentence, program)
+    let [spansBySentencePos, spansByProgramPos_sorted, sortIndices] = findSpanPositions(id, sentence, program);
```
Comment (gcampax) [inline tool/requote.js]: By the way, it's very jarring to mix snake_case and camelCase _in the same variable_. Please pick one or the other - preferrably camelCase unless it's a key JSON or a Thingpedia parameter.

Comment (Mehrad0711) [reply tool/requote.js]: Yeah, the pythonic way of thinking sn(e)aked in here...
